Lib/importlib/metadata.py

```diff
@@ -0,0 +1,271 @@
+        if not root.is_dir():
+            return ()
+        return (
+            item
+            for item in root.iterdir()
+            if item.is_dir()
+            and str(item.name).startswith(name)
+            and re.match(rf'{name}(-.*)?\.(dist|egg)-info', str(item.name))
+            )
+
+
+class PathDistribution(Distribution):
+    def __init__(self, path):
+        """Construct a distribution from a path to the metadata directory."""
+        self._path = path
+
+    def load_metadata(self, name):
+        """Attempt to load metadata given by the name.
+
+        :param name: The name of the distribution package.
+        :return: The metadata string if found, otherwise None.
+        """
+        filename = os.path.join(self._path, name)
+        with contextlib.suppress(FileNotFoundError):
+            with open(filename, encoding='utf-8') as fp:
+                return fp.read()
+        return None
+
+
+class WheelMetadataFinder:
+    """A degenerate finder for distribution packages in wheels.
+
+    This finder supplies only a find_distribution() method for versions
+    of Python that do not have a PathFinder find_distribution().
+    """
+    @staticmethod
+    def find_spec(*args, **kwargs):
+        return None
+
+    @classmethod
+    def find_distribution(cls, name):
+        try:
+            module = import_module(name)
+        except ImportError:
+            return None
+        archive = getattr(module.__loader__, 'archive', None)
+        if archive is None:
+            return None
+        try:
+            name, version = Path(archive).name.split('-')[0:2]
+        except ValueError:
+            return None
+        dist_info = '{}-{}.dist-info'.format(name, version)
+        with ZipFile(archive) as zf:
+            # Since we're opening the zip file anyway to see if there's a
+            # METADATA file in the .dist-info directory, we might as well
+            # read it and cache it here.
+            zi = zf.getinfo('{}/{}'.format(dist_info, 'METADATA'))
+            metadata = zf.read(zi).decode('utf-8')
+        return WheelDistribution(archive, dist_info, metadata)
+
+
+class WheelDistribution(Distribution):
+    def __init__(self, archive, dist_info, metadata):
+        self._archive = archive
+        self._dist_info = dist_info
+        self._metadata = metadata
+
+    def load_metadata(self, name):
+        """Attempt to load metadata given by the name.
+
+        :param name: The name of the distribution package.
+        :return: The metadata string if found, otherwise None.
+        """
+        if name == 'METADATA':
+            return self._metadata
+        with ZipFile(self._archive) as zf:
+            with contextlib.suppress(KeyError):
+                as_bytes = zf.read('{}/{}'.format(self._dist_info, name))
+                return as_bytes.decode('utf-8')
+        return None
+
+
+def _install():                                     # pragma: nocover
```

Comment:
Is this how Finders are added to `sys.metapath` in other cases? I guess I imagined when we ported this to 3.8 that there would be a more direct creation of these finders... and that in fact the normal PathFinder would supply find_distribution (superseding MetadataPathFinder and WheelMetadataPathFinder). But maybe this approach is suitable. Still pondering...

Reply:
I'm going to defer this to the next port attempt.